tf2_ros/src/tf2_ros/multiple_static_transform_broadcaster.py

```diff
@@ -0,0 +1,53 @@
+# Software License Agreement (BSD License)
+#
+# Copyright (c) 2008, Willow Garage, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above
+#    copyright notice, this list of conditions and the following
+#    disclaimer in the documentation and/or other materials provided
+#    with the distribution.
+#  * Neither the name of the Willow Garage nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import rospy
+from tf2_msgs.msg import TFMessage
+
+
+class MultipleStaticTransformBroadcaster(object):
+    """
+    :class:`MultipleStaticTransformBroadcaster`
+    keeps track of all unique transforms and resends them all to new subscribers
+    """
+    # Since only one latched publisher should exist per node, this is a class object
+    pub_tf = rospy.Publisher("/tf_static", TFMessage, queue_size=100, latch=True)
+    tfs = dict()
+
+    @classmethod
+    def sendTransform(cls, transform):
+        if not isinstance(transform, list):
+            transform = [transform]
+        for tf in transform:
+            cls.tfs[(tf.header.frame_id, tf.child_frame_id)] = tf
+
+        cls.pub_tf.publish(TFMessage(cls.tfs.values()))
```

Comment:
```suggestion
        cls.pub_tf.publish(TFMessage(list(cls.tfs.values())))
```

Without this change, TFMessage constructor fails on Python 3 as values() returns not directly a list, but a proxy object.

Reply:
What do you mean? We have this code in use for a year now on python3. What fails?

Reply:
Without this fix, as soon as there is a subscriber to /tf_static, the publisher starts yelling:

    Inbound TCP/IP connection failed: 'dict_values' object has no attribute 'header'

This happens on Noetic/Focal to me on several machines. On Python 3, `type({1: 2}.values())` yields type `dict_values`.

Without /tf_static subscribers, this error does not manifest.

You can try it yourself with this test node (rewrite it to use your mutli tf publisher): https://github.com/ctu-vras/ros-utils/blob/master/cras_py_common/test/publish_static_tfs.py . Run this node in one console and `rosrun tf tf_echo a c` in another one.

---

Melodic/Bionic does not have this problem becuase `type({1: 2}.values())` is directly `list`.